app/controllers/courses/sites_controller.rb

```diff
@@ -8,6 +8,30 @@ class SitesController < ApplicationController
 
     def edit; end
 
+    def update
+      @course.provider_code = @provider.provider_code
+      @course.sites = params.dig(:course, :site_statuses_attributes)
+        &.values
+        &.select { |field| field["selected"] == "1" }
+        &.map { |field| @provider.sites.find { |site| site.id == field["id"] } }
+
+      # If I call @course.save directly, and the API returns an error,
+      # it seems to mutate the @course object, removing `site_statuses`
+      # and causing `render :edit` to fail.
+      no_idea_why_i_have_to_dup_this_course = @course.dup
```

Comment:
I don't want to merge this, obviously, but I can't figure out why I can't call `@course.save` directly...

Reply:
This is probably because the API doesn't allow attributes it isn't expecting?

Reply:
It sounds like the question is why is `@course` changing at all when an error is returned.

Shot in the dark, but could this be related to the through relationship? I can see how it might be a little be confused if you modify `sites` here and then when it tries to save it to the backend it gets an error ... then is `site_statuses` still valid since you've just tried to save a change to it and it's come back with an error? I'd have to play around with it to understand it better.

Reply:
Is this related to the work you were doing recently to remove the `site_statuses` from the response sent by the backend?

Reply:
Sorted it out, seems to work without this hack now.